internal/mining/mining_view.go

```diff
@@ -0,0 +1,464 @@
+// Copyright (c) 2020 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mining
+
+import "github.com/decred/dcrd/chaincfg/chainhash"
+
+const (
+	// AncestorTrackingLimit is the maximum number of ancestors a transaction
+	// can have ancestor stats calculated for.
+	AncestorTrackingLimit = 25
+)
+
+// defaultAncestorStats is the default-state of ancestor stats for a transaction
+// that has no ancestors.
+var defaultAncestorStats = &TxAncestorStats{}
+
+// TxMiningView represents a snapshot of all transactions in mempool
+// as well as the hierarchy of transactions, as they may spend from one another
+// in the mempool.
+type TxMiningView struct {
+	rejected           map[chainhash.Hash]struct{}
+	txGraph            *txDescGraph
+	txDescs            []*TxDesc
+	trackAncestorStats bool
+	ancestorStats      map[chainhash.Hash]*TxAncestorStats
+}
+
+// NewTxMiningView creates a new mining view instance.  The forEachRedeemer
+// parameter should define the function to be used for finding transactions that
+// spend a given transaction.
+func NewTxMiningView(enableAncestorTracking bool, forEachRedeemer forEachRedeemer) *TxMiningView {
+	return &TxMiningView{
+		rejected:           make(map[chainhash.Hash]struct{}),
+		txGraph:            newTxDescGraph(forEachRedeemer),
+		txDescs:            nil,
+		trackAncestorStats: enableAncestorTracking,
+		ancestorStats:      make(map[chainhash.Hash]*TxAncestorStats),
+	}
+}
+
+// addAncestorTo modifies the TxAncestorStats instance to include the provided
+// TxDesc's statistics.
+func addAncestorTo(stats *TxAncestorStats, txDesc *TxDesc) {
+	stats.Fees += txDesc.Fee
+	stats.SizeBytes += txDesc.TxSize
+	stats.TotalSigOps += txDesc.TotalSigOps
+	stats.NumAncestors++
+}
+
+// addDescendantTo modifies the TxAncestorStats instance to account for the
+// addition of a newly tracked descendant.
+func addDescendantTo(stats *TxAncestorStats) {
+	stats.NumDescendants++
+}
+
+// removeAncestorFrom modifies the TxAncestorStats instance to stop storing
+// the statistics of the provided TxDesc.
+func removeAncestorFrom(stats *TxAncestorStats, txDesc *TxDesc) {
+	stats.Fees -= txDesc.Fee
+	stats.SizeBytes -= txDesc.TxSize
+	stats.TotalSigOps -= txDesc.TotalSigOps
+	stats.NumAncestors--
+}
+
+// removeDescendantFrom modifies the TxAncestorStats instance to account for
+// the removal of a descendant.
+func removeDescendantFrom(stats *TxAncestorStats) {
+	stats.NumDescendants--
+}
+
+// Ancestors returns a collection of all transactions in the graph that the
+// provided transaction hash depends on, and its ancestors' bundle stats.
```

Comment:
Since this is exported now, I think it should specifically call out that it is NOT safe for concurrent access.  That same goes for the other exported methods (unless of course they actually are).  That might not be obvious otherwise due to the fact there is a `Clone`.

Reply:
Good call, I updated this and the other exported methods to call out that they are not safe for concurrent access.